File: repo/react/markpad/src/lib/session.ts
// Single chokepoint for session persistence (load/save the recents list + active state).
// The on-disk session.json is owned by the Rust side (src-tauri/src/session.rs); this module
// is the only TS importer of `invoke("load_session" | "save_session")`. The schema mirrors
// the Rust SessionRecord struct exactly (snake_case fields, version=2).
//
// Schema v2 (recents redesign): items replace tabs. A clean file item stores only its path;
// modified and untitled items also carry their unsaved buffer contents (text/saved_text) so
// drafts survive a restart. The Rust side migrates legacy v1 records transparently, so this
// module always receives a v2 record.

import { invoke } from "@tauri-apps/api/core";

export type SessionItem = {
  kind: "file" | "untitled";
  /** Absolute path for file items; null for untitled drafts. */
  path: string | null;
  /** Display name (basename for files, "Untitled-N" for drafts). */
  name: string;
  /** Whether the item had unsaved changes when persisted. */
  dirty: boolean;
  /** Unsaved buffer text — present for retained (untitled/dirty) items, else null. */
  text: string | null;
  /** Disk baseline text captured alongside `text` for accurate dirty tracking. */
  saved_text: string | null;
  /** Manual language override ("markdown" | "json" | "yaml"); null when the
      language derives from the path. Absent in records written before JSON
      support — the Rust side defaults it to null. */
  language: string | null;
};

export type SessionRecord = {
  version: 2;
  items: SessionItem[];
  active_index: number | null;
};

function emptySession(): SessionRecord {
  return { version: 2, items: [], active_index: null };
}

export async function loadSession(): Promise<SessionRecord> {
  try {
    const record = await invoke<SessionRecord>("load_session");
    if (record && record.version === 2 && Array.isArray(record.items)) {
      return record;
    }
    return emptySession();
  } catch (err) {
    console.warn("load_session failed; treating as empty session:", err);
    return emptySession();
  }
}

export async function saveSession(record: SessionRecord): Promise<void> {
  try {
    await invoke<void>("save_session", { record });
  } catch (err) {
    console.warn("save_session failed; session not persisted this round:", err);
  }
}

File: repo/react/markpad/src-tauri/src/session.rs
use serde::{Deserialize, Serialize};
use tauri::Manager;

// Session schema v2 (Feature: recents redesign).
//
// The left "recents" list persists up to 50 most-recently-opened items. Unlike
// v1 (which stored only paths and reloaded everything from disk), v2 also
// carries the unsaved buffer contents of modified and untitled items so drafts
// survive a restart. Clean file items still store only their path and are
// reloaded lazily from disk on the TS side.
//
//   kind == "file"     -> `path` is Some; a clean file omits text/saved_text.
//   kind == "untitled" -> `path` is None; text/saved_text carry the draft.
//
// v1 records (`{version:1, tabs:[{path}], active_index}`) are migrated on load
// so an upgrade doesn't lose the user's open files.

#[derive(Serialize, Deserialize)]
pub struct SessionRecord {
    pub version: u32,
    pub items: Vec<SessionItem>,
    pub active_index: Option<usize>,
}

#[derive(Serialize, Deserialize)]
pub struct SessionItem {
    pub kind: String,
    #[serde(default)]
    pub path: Option<String>,
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub dirty: bool,
    #[serde(default)]
    pub text: Option<String>,
    #[serde(default)]
    pub saved_text: Option<String>,
    /// Manual language override ("markdown" | "json" | "yaml"); None when the
    /// language derives from the path. The value is opaque here — the frontend
    /// validates it (see asDocumentLanguage), so a record written by a newer
    /// version loads unchanged. Defaults keep pre-JSON-support records loadable.
    #[serde(default)]
    pub language: Option<String>,
}

// --- v1 (legacy) shape, kept only to migrate old session files. ---

#[derive(Deserialize)]
struct SessionRecordV1 {
    version: u32,
    #[serde(default)]
    tabs: Vec<SessionTabEntryV1>,
    #[serde(default)]
    active_index: Option<usize>,
}

#[derive(Deserialize)]
struct SessionTabEntryV1 {
    path: String,
}

fn empty_session() -> SessionRecord {
    SessionRecord {
        version: 2,
        items: Vec::new(),
        active_index: None,
    }
}

fn basename(path: &str) -> String {
    let normalized = path.replace('\\', "/");
    match normalized.rsplit('/').next() {
        Some(name) if !name.is_empty() => name.to_string(),
        _ => normalized,
    }
}

fn migrate_v1(v1: SessionRecordV1) -> SessionRecord {
    let items = v1
        .tabs
        .into_iter()
        .map(|tab| SessionItem {
            name: basename(&tab.path),
            path: Some(tab.path),
            kind: "file".to_string(),
            dirty: false,
            text: None,
            saved_text: None,
            language: None,
        })
        .collect();
    SessionRecord {
        version: 2,
        items,
        active_index: v1.active_index,
    }
}

fn parse_session(content: &str) -> SessionRecord {
    if let Ok(record) = serde_json::from_str::<SessionRecord>(content) {
        if record.version == 2 {
            return record;
        }
    }
    if let Ok(v1) = serde_json::from_str::<SessionRecordV1>(content) {
        if v1.version == 1 {
            return migrate_v1(v1);
        }
    }
    empty_session()
}

#[tauri::command]
pub async fn load_session(app: tauri::AppHandle) -> Result<SessionRecord, String> {
    let dir = app
        .path()
        .app_data_dir()
        .map_err(|err| format!("failed to resolve app_data_dir: {err}"))?;
    let path = dir.join("session.json");
    let content = match std::fs::read_to_string(&path) {
        Ok(s) => s,
        Err(_) => return Ok(empty_session()),
    };
    Ok(parse_session(&content))
}

#[tauri::command]
pub async fn save_session(app: tauri::AppHandle, record: SessionRecord) -> Result<(), String> {
    let dir = app
        .path()
        .app_data_dir()
        .map_err(|err| format!("failed to resolve app_data_dir: {err}"))?;
    std::fs::create_dir_all(&dir).map_err(|err| format!("failed to create app_data_dir: {err}"))?;
    let serialized = serde_json::to_string_pretty(&record)
        .map_err(|err| format!("failed to serialize session: {err}"))?;
    let tmp = dir.join("session.json.tmp");
    let final_path = dir.join("session.json");
    std::fs::write(&tmp, serialized)
        .map_err(|err| format!("failed to write session.json.tmp: {err}"))?;
    std::fs::rename(&tmp, &final_path)
        .map_err(|err| format!("failed to rename session.json.tmp: {err}"))?;
    Ok(())
}
